pmd-java/src/main/java/net/sourceforge/pmd/lang/java/oom/Metrics.java

```diff
@@ -53,7 +54,7 @@ private Metrics() { // Cannot be instantiated
      *
      * @return The value of the metric, or {@code Double.NaN} if the value couln't be computed
      */
-    public static double get(ClassMetricKey key, ASTAnyTypeDeclaration node) {
+    public static double get(MetricKey<ClassMetric> key, ASTAnyTypeDeclaration node) {
```

Comment:
The use of a `static` singleton pattern is concerning to me.  I've come to see it as an antiquated style which causes _far_ _more_ problems than it's worth.  I find it generally better to code cleanly without statics, and then directly address the scoping problem _outside_ of the implementation.  Thus, if PMD needs to use a _single_ instance of this state for some scope of processing, or we need Rules to have convenient access, then directly address those problems.  It's not the implementation's issue (i.e. metrics).

See `PMDConfiguration` and `RuleContext` as existing mechanisms for the propagation/scoping of state, and the high level coordination performed by `MultiThreadProcessor` and `PmdRunnable`.  PMD juggles quite a bit of complex state internally, from multi-threaded source file processing to reporting rendering.

It's also usually a pain to unit test `static` stuff... mocking is harder, you have to be _very_ careful to cleanup state between tests, etc.